### PR TITLE
Convolution logic (part 3)

### PR DESCRIPTION
**Pull request by bernardkwok**
- Make it so we use nodegroup to indicate type of nodes which can be samplers and those that are can be sampled. Final names as discussed can gone over with ILM .
-  For now recognized "texcoord" and "position" as default inputs which can be sampled. We can determine how to set additional meta data (whether it's "defaultgeomgroup" or other) for tagging.
- Extract out sampling utilities which are adding is there are samplers in the graph ("convolutions" currently).
- Make SgNodes keep track of a sample input so no scanning is required anymore.

This is set up for transition of logic to for sample generation away from the sampler.

### CODE REVIEW

File: source/MaterialXGenGlsl/Nodes/HeightToNormalGlsl.cpp [hunk 1]
```diff
@@ -18,53 +18,6 @@ namespace MaterialX
         shader.createUniform(HwShader::PIXEL_STAGE, HwShader::PUBLIC_UNIFORMS, Type::VECTOR2, "u_screenSize" /*, OGSFX_SIZE_SEMANTIC*/);
     }
 
-    void HeightToNormalGlsl::emitFunctionDefinition(const SgNode& /*node*/, ShaderGenerator& shadergen_, Shader& shader_)
```
Comment (bernardkwok) [inline source/MaterialXGenGlsl/Nodes/HeightToNormalGlsl.cpp]: This is all in an include file now so it's easier to modify and add to.

File: source/MaterialXGenGlsl/Nodes/HeightToNormalGlsl.cpp [hunk 2]
```diff
@@ -87,10 +40,15 @@ namespace MaterialX
                     {
                         string outputName = upstreamOutput->name;
 
-                        // The only sample input to consider now are texcoord inputs .
-                        SgInput* sampleInput = upstreamNode->getInput("texcoord");
+                        // Find out which input needs to be sampled multiple times
+                        SgInput* samplingInput = upstreamNode->getSamplingInput();
```
Comment (bernardkwok) [inline source/MaterialXGenGlsl/Nodes/HeightToNormalGlsl.cpp]: Get's the input which can be sampled if any.

File: source/MaterialXGenShader/SgNode.cpp [hunk 3]
```diff
@@ -355,6 +378,18 @@ void SgNode::renameOutput(const string& name, const string& newName)
     }
 }
 
+bool SgNode::elementCanBeSampled2D(const Element& element) const
```
Comment (bernardkwok) [inline source/MaterialXGenShader/SgNode.cpp]: Default what can be sampled, can extend later on if we have otherways to determine this.

File: source/MaterialXGenShader/SgNode.h [hunk 4]
```diff
@@ -200,22 +212,58 @@ class SgNode
     void renameInput(const string& name, const string& newName);
     void renameOutput(const string& name, const string& newName);
 
+    /// Get input used for sampling. May not be any
+    SgInput* getSamplingInput() const
+    {
+        return _samplingInput;
+    }
+
+    /// Check if a given node can be sampled in 2D (uv space)
```
Comment (bernardkwok) [inline source/MaterialXGenShader/SgNode.h]: utils to tell what can be sampled and what is a sampler

**Comment by niklasharrysson** [comment]
As a general comment, I feel the changes to SgNode here is a move in the wrong direction. It would be better to keep SgNode general and have these sampling/filtering features handled by the SgImplementation classes, since this is specific to different node types. It could be handled by specific Texture2D, Texture3D classes derived from class SourceCode.

Maybe that is the intention, since we discussed that earlier, and these are temporary changes to try things out?

**Pull request closed without merging**